File: 킬링캠프/3주 차 - BFS, DFS/challenge/거리두기 확인하기/Java/DX, DY (깡구현)/Solution.java
class Solution {

    private static final int[][] DIR = new int[][]{{-1, 0}, {0, 1}, {0, -1}, {1, 0}};   // 북, 동, 서, 남

    public int[] solution(String[][] places) {
        int[] result = new int[places.length];
        outer: for (int i = 0; i < places.length; i++) {
            char[][] target = createTarget(places[i]);
            for (int row = 0; row < target.length; row++) {
                for (int col = 0; col < target[row].length; col++) {
                    if (target[row][col] != 'P') {
                        continue;
                    }

                    if (!check(target, row, col)) {
                        continue outer;
                    }
                }
            }

            result[i] = 1;
        }

        return result;
    }

    private char[][] createTarget(String[] place) {
        char[][] target = new char[5][5];
        for (int i = 0; i < 5; i++) {
            String s = place[i];
            for (int j = 0; j < 5; j++) {
                target[i][j] = s.charAt(j);
            }
        }

        return target;
    }

    private boolean check(char[][] places, int startRow, int startCol) {
        for (int i = 0; i < DIR.length; i++) {
            int oneStepRow = startRow + DIR[i][0];
            int oneStepCol = startCol + DIR[i][1];

            if (!isSafe(places, oneStepRow, oneStepCol)) {
                continue;
            }

            if (places[oneStepRow][oneStepCol] == 'P') {
                return false;
            }

            for (int j = 0; j < DIR.length; j++) {
                if (j == (DIR.length - i - 1)) {
                    continue;
                }

                int twoStepRow = oneStepRow + DIR[j][0];
                int twoStepCol = oneStepCol + DIR[j][1];

                if (!isSafe(places, twoStepRow, twoStepCol)) {
                    continue;
                }

                if (places[oneStepRow][oneStepCol] != 'X' && places[twoStepRow][twoStepCol] == 'P') {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean isSafe(char[][] places, int row, int col) {
        return row >= 0 && row < places.length && col >= 0 && col < places[row].length;
    }
}

File: 킬링캠프/3주 차 - BFS, DFS/challenge/거리두기 확인하기/DX, DY (깡구현)/solution.py
DIR = [[-1, 0], [0, 1], [0, -1], [1, 0]]

def solution(places):
    answer = [1 for _ in range(5)]
    for i in range(5):
        if not has_distance(places[i]):
            answer[i] = 0
    
    return answer

def has_distance(place):
    for row in range(5):
        for col in range(5):
            if place[row][col] != "P":
                continue

            if not check_distance(place, row, col):
                return False
    
    return True

def check_distance(place, row, col):
    for one_step_d in range(len(DIR)):
        one_step_row = row + DIR[one_step_d][0]
        one_step_col = col + DIR[one_step_d][1]

        if not is_safe(one_step_row, one_step_col):
            continue

        if place[one_step_row][one_step_col] == "P":
            return False
        elif place[one_step_row][one_step_col] == "X":
            continue
        
        for two_step_d in range(len(DIR)):
            if two_step_d == len(DIR) - one_step_d - 1:
                continue

            two_step_row = one_step_row + DIR[two_step_d][0]
            two_step_col = one_step_col + DIR[two_step_d][1]
            if not is_safe(two_step_row, two_step_col):
                continue

            if place[two_step_row][two_step_col] == "P":
                return False
    
    return True

def is_safe(row, col):
    return row >= 0 and row < 5 and col >= 0 and col < 5
